### PR TITLE
codegen: Add support for list and maps generated as value members instead of pointer

### PR DESCRIPTION
**Pull request by jasdel**
Updates the SDK code generation to make use of the new smithy `NullableIndex` and smithy-go's `GoPointableIndex`. These utilities allow the SDK's code generation to be aware of API shapes and members that are nullable, and can be rendered as values by the code generation instead of pointers.

This update updates the SDK's code generation types to be value instead of pointer in most cases.
* Numbers (int8, int16, int32, int64, float32, float64)
* bool
* lists - e.g. `[]string` instead of `[]*string`
* maps - e.g. `map[string]string` instead of `map[string]*string`.

Number and boolean types that are not boxed (not pointer) will not serialize their zero value unless they are members of a list or map.

This change also significantly refactors how the SDK represents nullability and pointability when generation types and members from an API model. Utilities were added in the smithy-go change that help ensure more consistent logic through the sdk's code generation.

Depends on https://github.com/awslabs/smithy-go/pull/232
Fixes https://github.com/aws/aws-sdk-go-v2/issues/896

### CODE REVIEW

**Review by skmcgrail — APPROVED** [review]

**Comment by skmcgrail** [comment]
Have you ran `make integration` to verify these changes?

**Comment by jasdel** [comment]
Yeah verified that all integration tests pass successfully.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java [hunk 1]
```diff
@@ -48,14 +53,16 @@
...
+            boolean isXmlAttributeMember
     ) {
         this.context = context;
+        this.member = member;
         this.dataDest = dataDest;
         this.timestampFormat = timestampFormat;
         this.isXmlAttributeMember = isXmlAttributeMember;
-        this.isFlattened = isFlattened;
+        this.isFlattened = member.hasTrait(XmlFlattenedTrait.ID);
+        ;
```
Comment (skotambkar) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java]: accidental?

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java [hunk 2]
```diff
@@ -69,21 +71,21 @@ public XmlShapeDeserVisitor(GenerationContext context, Predicate<MemberShape> me
...
 
     // generates code to define and initialize output variable for an aggregate shape
-    private void generatesIntializerForOutputVariable(GenerationContext context, Shape shape) {
+    private void generatesInitializerForOutputVariable(GenerationContext context, Shape shape) {
         GoWriter writer = context.getWriter();
         Symbol shapeSymbol = context.getSymbolProvider().toSymbol(shape);
 
         writer.write("var sv $P", shapeSymbol);
+        // TODO [denseListMap] need to reference if pointable type. Assumes containers are always double pointers.
+        // TODO this probably isn't true for union types.
```
Comment (skotambkar) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java:88]: still todo's?

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java [hunk 3]
```diff
@@ -48,14 +53,16 @@
     private final boolean isFlattened;
 
     public XmlMemberDeserVisitor(
-            GenerationContext context, String dataDest, Format timestampFormat,
-            boolean isXmlAttributeMember, boolean isFlattened
+            GenerationContext context, MemberShape member, String dataDest, Format timestampFormat,
+            boolean isXmlAttributeMember
```
Comment (skotambkar) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java:57]: Nit: one on each line

**Review by skotambkar — APPROVED** [review]

**Comment by skotambkar** [comment]
Looks like smithy-go dep version is not correct? CI for Java fails as it doesn't find methods from awslabs/smithy-go#232

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java (hunk 2 above)
Comment (jasdel) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java:88]: I thought I removed all of those, must of missed one, thanks!

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java (hunk 1 above)
Comment (jasdel) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlMemberDeserVisitor.java]: Yeah typo

**Comment by jasdel** [comment]
That's probably result of bad merge. There where no Go changes in the smithy-go change. I won't be able to have CI succeed on this PR until the smithy change is merged because it depends on Java changes in that pr.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java (hunk 2 above)
Comment (jasdel) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java:88]: This comment was wrong, union de-serialization will always be a struct that satisfy the union member interface. And all types in this context will be pointers like list/maps, or double pointers like structs.

**Pull request merged**